src/Microsoft.Health.Fhir.SqlServer/Features/Search/HashingSqlQueryParameterManager.cs

```diff
@@ -0,0 +1,201 @@
+                        break;
+                    case SqlDbType.Int:
+                        WriteAndAdvance(buf, ref currentBufferIndex, ref incrementalHash, (int)sqlParameter.Value);
+                        break;
+                    case SqlDbType.SmallInt:
+                        WriteAndAdvance(buf, ref currentBufferIndex, ref incrementalHash, (short)sqlParameter.Value);
+                        break;
+                    case SqlDbType.TinyInt:
+                        WriteAndAdvance(buf, ref currentBufferIndex, ref incrementalHash, (byte)sqlParameter.Value);
+                        break;
+                    case SqlDbType.UniqueIdentifier:
+                        WriteAndAdvance(buf, ref currentBufferIndex, ref incrementalHash, (Guid)sqlParameter.Value);
+                        break;
+                    case SqlDbType.NChar:
+                    case SqlDbType.NText:
+                    case SqlDbType.VarChar:
+                    case SqlDbType.NVarChar:
+                    case SqlDbType.Text:
+                        WriteAndAdvanceString(buf, ref currentBufferIndex, ref incrementalHash, (string)sqlParameter.Value);
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unexpected parameter type {sqlParameter.SqlDbType}");
+                }
+            }
+
+            Span<byte> hashBytes = stackalloc byte[32];
+
+            if (incrementalHash != null)
+            {
+                if (currentBufferIndex > 0)
+                {
+                    incrementalHash.AppendData(buf[..currentBufferIndex]);
+                }
+
+                incrementalHash.GetCurrentHash(hashBytes);
+                incrementalHash.Dispose();
+            }
+            else
+            {
+                if (!SHA256.TryHashData(buf[..currentBufferIndex], hashBytes, out _))
+                {
+                    throw new InvalidOperationException("Failed to hash data");
+                }
+            }
+
+            Span<char> hashChars = stackalloc char[44]; // 44 since inputLength = 32 and inputLength => (inputLength / 3 * 4) + (((inputLength % 3) != 0) ? 4 : 0) = 44
+
+            if (!Convert.TryToBase64Chars(hashBytes, hashChars, out int hashCharsLength))
+            {
+                throw new InvalidOperationException("Failed to convert to Base64 chars.");
+            }
+
+            stringBuilder.Append(hashChars[..hashCharsLength]);
+        }
+
+        private static void WriteAndAdvance<T>(Span<byte> buffer, ref int currentIndex, ref IncrementalHash incrementalHash, T element)
+            where T : struct
+        {
+            int elementLength = Unsafe.SizeOf<T>();
+            if (currentIndex + elementLength > buffer.Length)
+            {
+                incrementalHash ??= IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
```

Comment:
>HashAlgorithmName.SHA256 [](start = 63, length = 24)

nit: since security is not important here, would it be helpful to use a faster non-cryptographic hash algorithm? I guess the length of the data to hash is small enough here that it shouldn't really have an impact.

Reply:
Yes, this is a good point. I tried out xxHash using [this ](https://github.com/uranium62/xxHash) implementation.

With parameters `@p0=690,@p1='1982-11-05 00:00:00',@p2='1982-11-05 23:59:59.9999999',@p3=103,@p4=1016,@p5=N'Smith%',@p6=11`, here was the difference:

| Method |       Mean |    Error |   StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |-----------:|---------:|---------:|------:|------:|------:|----------:|
|      Sha256 | 1,155.3 ns | 22.65 ns | 35.92 ns |     - |     - |     - |         - |
|      xxHash |   268.6 ns |  5.30 ns |  4.96 ns |     - |     - |     - |         - |

But I'd like to hold off on making this change though, until we evaluate the package.